### PR TITLE
dump and restore cpu affinity of each thread

### PR DESCRIPTION
**Pull request by hdzhoujie**
dump and restore cpu affinity of each thread.

Add one entry of thread_cpuallow_entry into
thread_core_entry to save cpu affinity info.

Restore it after threads restored but before running.

Add option --with-cpu-affinity to enable this function at restore.
Signed-off-by: zhou jie <zhoujie133@huawei.com>
Signed-off-by: He jingxian <hejingxian@huawei.com>
Signed-off-by: Sang Yan <sangyan@huawei.com>

### CODE REVIEW

**Review by hdzhoujie — COMMENTED** [review]
dump and restore cpu affinity of each thread.

Add one entry of thread_cpuallow_entry into
thread_core_entry to save cpu affinity info.

Restore it after threads restored but before running.

Add option --with-cpu-affinity to enable this function at restore.
Signed-off-by: zhou jie [zhoujie133@huawei.com](mailto:zhoujie133@huawei.com)
Signed-off-by: He jingxian [hejingxian@huawei.com](mailto:hejingxian@huawei.com)
Signed-off-by: Sang Yan [sangyan@huawei.com](mailto:sangyan@huawei.com)

File: criu/pie/restorer.c [hunk 1]
```diff
@@ -425,6 +425,32 @@ static int restore_signals(siginfo_t *ptr, int nr, bool group)
...
 
+static int restore_cpu_affinity(struct task_restore_args *args)
+{
+	int i;
+	int pid;
+	int ret;
+	cpu_set_t *cpumask;
+	cpu_set_t *allowed_cpus;
+
+	if (!args->with_cpu_affinity) {
```
Comment (minhbq-99) [inline criu/pie/restorer.c:436]: Is it a problem in case in dump we don't use `with_cpu_affinity` but in restore we use `with_cpu_affinity`? As far as I can see, the value of cpumask in core image is uninitialized when dumping without `with_cpu_affinity`.

File: criu/cr-restore.c [hunk 2]
```diff
@@ -3275,6 +3279,25 @@ static int prepare_signals(int pid, struct task_restore_args *ta, CoreEntry *lea
 	return ret;
 }
 
+static int prepare_allowed_cpus(int pid, struct task_restore_args *ta, CoreEntry *leader_core)
+{
+	int i;
+	cpu_set_t *cpumaks;
```
[the remark this answers is not in the document]
Comment (hdzhoujie) [reply criu/cr-restore.c]: thanks, I have modified it.

File: images/core.proto [hunk 3]
```diff
@@ -105,6 +109,7 @@ message thread_core_entry {
 	optional string			comm		= 13;
 	optional uint64			blk_sigset_extended	= 14;
 	optional rseq_entry		rseq_entry	= 15;
+	optional thread_allowedcpus_entry 	allowed_cpus	= 16;
```
[the remark this answers is not in the document]
Comment (hdzhoujie) [reply images/core.proto:113]: now thread_ allowedcpus_ Entry has two variables, has_cpumask is used for compatibility. In dump, we are not compatible with CPU_ Affinity is used together, but in restore, we use it together with cpu Affinity.

File: criu/pie/restorer.c (hunk 1 above)
Comment (hdzhoujie) [reply criu/pie/restorer.c:437]: thanks, I add has_cpumask variable solves this problem.

File: criu/include/restorer.h [hunk 4]
```diff
@@ -103,6 +104,8 @@ struct thread_restore_args {
 
 	struct task_restore_args *ta;
 
+	char *allowed_cpus;
```
[the remark this answers is not in the document]
Comment (hdzhoujie) [reply criu/include/restorer.h]: thanks, deleted.

**Comment by rst0git** [comment]
@hdzhoujie Would you be able to add a "Signed-off-by" line in the commit message of commit `dump and restore threads or processes cpu affinity`?

File: test/zdtm/static/cpu-affinity0.c [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+   test_init(argc, argv);
+
+   CPU_ZERO(&old);
+   CPU_ZERO(&new);
+   
+   /* test only 0 core because of CI test env limited */
+   CPU_SET(0, &old);
+
+   if (sched_setaffinity(getpid(), sizeof(old), &old) < 0) {
+       pr_perror("Can't set old cpu affinity! errno: %d", errno);
```
Comment (rst0git) [inline test/zdtm/static/cpu-affinity0.c:24]: `pr_perror` already prints the value of `errno`:
https://github.com/checkpoint-restore/criu/blob/b3a950bf7c7c286e781eb02230af21324853b4fb/criu/include/log.h#L61

File: test/zdtm/static/cpu-affinity0.c [hunk 6]
```diff
@@ -0,0 +1,42 @@
+#include <errno.h>
```
Comment (avagin) [inline test/zdtm/static/cpu-affinity0.c:1]: The test has to be in a separate patch.

**Comment by avagin** [comment]
How do you handle a  case when processes are restored on hosts with more or less cpus?

File: test/zdtm/static/cpu-affinity0.c (hunk 6 above)
Comment (hdzhoujie) [reply test/zdtm/static/cpu-affinity0.c:1]: Thank you for your reply.
Do I need to commit in two parts, submit the code patch first, and then submit the test patch when it is received?

**Comment by hdzhoujie** [comment]
>

**Comment by hdzhoujie** [comment]
> How do you handle a case when processes are restored on hosts with more or less cpus?
Thank you for your reply.
This feature is used for backup or recovery of processes on the same host，eg: native kernel upgrade, checkout/restore of processes in the native machine

**Comment by rst0git** [comment]
>> How do you handle a case when processes are restored on hosts with more or less cpus?

> This feature is used for checkout/restore of processes on the same host，eg: native kernel upgrade, checkout/restore of processes in the native machine

@hdzhoujie Would it make sense to exit with an error when restoring on a host with different number of CPUs?

**Comment by hdzhoujie** [comment]
Thans for you reply.
exit with an error when restoring on a host with different number of CPUs

> > > How do you handle a case when processes are restored on hosts with more or less cpus?
> 
> > This feature is used for checkout/restore of processes on the same host，eg: native kernel upgrade, checkout/restore of processes in the native machine
> 
> @hdzhoujie Would it make sense to exit with an error when restoring on a host with different number of CPUs?

**Comment by hdzhoujie** [comment]
> > > How do you handle a case when processes are restored on hosts with more or less cpus?
> 
> > This feature is used for checkout/restore of processes on the same host，eg: native kernel upgrade, checkout/restore of processes in the native machine
> 
> @hdzhoujie Would it make sense to exit with an error when restoring on a host with different number of CPUs?

thanks for your reply.
In the recovery process, the judgment of the number of hosts is currently increased. 
If the --with-cpu-affinit option is enabled and the number of hosts is different during recovery, the recovery will fail.
